In [1]:
%load_ext sql

In [2]:
%sql mysql+pymysql://root:YvoneWarui@localhost:3306/chinook

# Database Setup
- **Opened the chinook dataset using MySql Database**

# **UNDERSTANDING THE DATABASE SCHEMA**
# **1. Showing the list of tables in the Chinook database**

In [4]:
%%sql
SHOW TABLES;

 * mysql+pymysql://root:***@localhost:3306/chinook
11 rows affected.


Tables_in_chinook
album
artist
customer
employee
genre
invoice
invoiceline
mediatype
playlist
playlisttrack


# **2. Inspecting the table structure for the 11 tables**

In [9]:
%%sql
DESCRIBE album;

 * mysql+pymysql://root:***@localhost:3306/chinook
3 rows affected.


Field,Type,Null,Key,Default,Extra
AlbumId,int,NO,PRI,None,
Title,varchar(160),NO,,None,
ArtistId,int,NO,MUL,None,


In [10]:
%%sql
DESCRIBE artist;

 * mysql+pymysql://root:***@localhost:3306/chinook
2 rows affected.


Field,Type,Null,Key,Default,Extra
ArtistId,int,NO,PRI,None,
Name,varchar(120),YES,,None,


In [11]:
%%sql
DESCRIBE customer;

 * mysql+pymysql://root:***@localhost:3306/chinook
13 rows affected.


Field,Type,Null,Key,Default,Extra
CustomerId,int,NO,PRI,None,
FirstName,varchar(40),NO,,None,
LastName,varchar(20),NO,,None,
Company,varchar(80),YES,,None,
Address,varchar(70),YES,,None,
City,varchar(40),YES,,None,
State,varchar(40),YES,,None,
Country,varchar(40),YES,,None,
PostalCode,varchar(10),YES,,None,
Phone,varchar(24),YES,,None,


In [12]:
%%sql
DESCRIBE employee;

 * mysql+pymysql://root:***@localhost:3306/chinook
15 rows affected.


Field,Type,Null,Key,Default,Extra
EmployeeId,int,NO,PRI,None,
LastName,varchar(20),NO,,None,
FirstName,varchar(20),NO,,None,
Title,varchar(30),YES,,None,
ReportsTo,int,YES,MUL,None,
BirthDate,datetime,YES,,None,
HireDate,datetime,YES,,None,
Address,varchar(70),YES,,None,
City,varchar(40),YES,,None,
State,varchar(40),YES,,None,


In [13]:
%%sql
DESCRIBE genre;

 * mysql+pymysql://root:***@localhost:3306/chinook
2 rows affected.


Field,Type,Null,Key,Default,Extra
GenreId,int,NO,PRI,None,
Name,varchar(120),YES,,None,


In [14]:
%%sql
DESCRIBE invoice;

 * mysql+pymysql://root:***@localhost:3306/chinook
9 rows affected.


Field,Type,Null,Key,Default,Extra
InvoiceId,int,NO,PRI,None,
CustomerId,int,NO,MUL,None,
InvoiceDate,datetime,NO,,None,
BillingAddress,varchar(70),YES,,None,
BillingCity,varchar(40),YES,,None,
BillingState,varchar(40),YES,,None,
BillingCountry,varchar(40),YES,,None,
BillingPostalCode,varchar(10),YES,,None,
Total,"decimal(10,2)",NO,,None,


In [15]:
%%sql
DESCRIBE invoiceline;

 * mysql+pymysql://root:***@localhost:3306/chinook
5 rows affected.


Field,Type,Null,Key,Default,Extra
InvoiceLineId,int,NO,PRI,None,
InvoiceId,int,NO,MUL,None,
TrackId,int,NO,MUL,None,
UnitPrice,"decimal(10,2)",NO,,None,
Quantity,int,NO,,None,


In [16]:
%%sql
DESCRIBE mediatype;

 * mysql+pymysql://root:***@localhost:3306/chinook
2 rows affected.


Field,Type,Null,Key,Default,Extra
MediaTypeId,int,NO,PRI,None,
Name,varchar(120),YES,,None,


In [17]:
%%sql
DESCRIBE playlist;

 * mysql+pymysql://root:***@localhost:3306/chinook
2 rows affected.


Field,Type,Null,Key,Default,Extra
PlaylistId,int,NO,PRI,None,
Name,varchar(120),YES,,None,


In [18]:
%%sql
DESCRIBE playlisttrack;

 * mysql+pymysql://root:***@localhost:3306/chinook
2 rows affected.


Field,Type,Null,Key,Default,Extra
PlaylistId,int,NO,PRI,None,
TrackId,int,NO,PRI,None,


In [19]:
%%sql
DESCRIBE track;

 * mysql+pymysql://root:***@localhost:3306/chinook
9 rows affected.


Field,Type,Null,Key,Default,Extra
TrackId,int,NO,PRI,None,
Name,varchar(200),NO,,None,
AlbumId,int,YES,MUL,None,
MediaTypeId,int,NO,MUL,None,
GenreId,int,YES,MUL,None,
Composer,varchar(220),YES,,None,
Milliseconds,int,NO,,None,
Bytes,int,YES,,None,
UnitPrice,"decimal(10,2)",NO,,None,


# **3. Generating a list of foreign keys(child table) where they are referenced from in the parent table(primary key)**

In [21]:
%%sql
SELECT 
    CONSTRAINT_NAME AS 'Constraint Name',
    TABLE_NAME AS 'Child Table (Foreign Key)',
    COLUMN_NAME AS 'Child Column',
    REFERENCED_TABLE_NAME AS 'Parent Table (Primary Key)',
    REFERENCED_COLUMN_NAME AS 'Parent Column'
FROM
    information_schema.KEY_COLUMN_USAGE
WHERE
    TABLE_SCHEMA = 'chinook' 
    AND REFERENCED_TABLE_NAME IS NOT NULL;

 * mysql+pymysql://root:***@localhost:3306/chinook
11 rows affected.


Constraint Name,Child Table (Foreign Key),Child Column,Parent Table (Primary Key),Parent Column
FK_AlbumArtistId,album,ArtistId,artist,ArtistId
FK_CustomerSupportRepId,customer,SupportRepId,employee,EmployeeId
FK_EmployeeReportsTo,employee,ReportsTo,employee,EmployeeId
FK_InvoiceCustomerId,invoice,CustomerId,customer,CustomerId
FK_InvoiceLineInvoiceId,invoiceline,InvoiceId,invoice,InvoiceId
FK_InvoiceLineTrackId,invoiceline,TrackId,track,TrackId
FK_PlaylistTrackPlaylistId,playlisttrack,PlaylistId,playlist,PlaylistId
FK_PlaylistTrackTrackId,playlisttrack,TrackId,track,TrackId
FK_TrackAlbumId,track,AlbumId,album,AlbumId
FK_TrackGenreId,track,GenreId,genre,GenreId


# **4. Counting the number of rows and columns from the tables**
- used **JOIN** to join two schemas i.e the TABLES and COLUMN schema to get the row count and the column count

In [24]:
%%sql
SELECT 
    t.TABLE_NAME AS 'Table Name',
    t.TABLE_ROWS AS 'Row Count (approx)',
    COUNT(c.COLUMN_NAME) AS 'Column Count'
FROM 
    information_schema.TABLES t
JOIN 
    information_schema.COLUMNS c 
    ON t.TABLE_SCHEMA = c.TABLE_SCHEMA 
    AND t.TABLE_NAME = c.TABLE_NAME
WHERE 
    t.TABLE_SCHEMA = 'chinook'
GROUP BY 
    t.TABLE_NAME, 
    t.TABLE_ROWS
ORDER BY 
    t.TABLE_ROWS DESC;

 * mysql+pymysql://root:***@localhost:3306/chinook
11 rows affected.


Table Name,Row Count (approx),Column Count
playlisttrack,8715,2
track,3483,9
invoiceline,2240,5
invoice,412,9
album,347,3
artist,275,2
customer,59,13
genre,25,2
playlist,18,2
employee,8,15


# Writing SQL queries using SELECT,WHERE,ORDER BY,GROUP BY,HAVING

In [28]:
%%sql
show tables;

 * mysql+pymysql://root:***@localhost:3306/chinook
11 rows affected.


Tables_in_chinook
album
artist
customer
employee
genre
invoice
invoiceline
mediatype
playlist
playlisttrack


# 1: Filtering and Sorting (SELECT, WHERE, ORDER BY)

In [3]:
%%sql
    -- This query finds all tracks that belong to Album ID 1 that are longer than 4 minutes, sorted from longest to shortest.
SELECT 
    TrackID, 
    Name, 
    AlbumID, 
    Milliseconds
FROM 
    Track
WHERE 
    AlbumID = 1
    AND 
    Milliseconds > 240000
ORDER BY 
    Milliseconds DESC;

 * mysql+pymysql://root:***@localhost:3306/chinook
4 rows affected.


TrackID,Name,AlbumID,Milliseconds
1,For Those About To Rock (We Salute You),1,343719
14,Spellbound,1,270863
10,Evil Walks,1,263497
12,Breaking The Rules,1,263288


In [4]:
%%sql
    -- This query finds all tracks that have a UnitPrice of exactly $0.99 and are shorter than 3 minutes (180,000 milliseconds)   
SELECT 
    TrackID,
    Name,
    UnitPrice,
    Milliseconds
FROM 
    track
WHERE 
    UnitPrice = 0.99
    AND Milliseconds < 180000
ORDER BY
    milliseconds ASC
    LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/chinook
10 rows affected.


TrackID,Name,UnitPrice,Milliseconds
2461,É Uma Partida De Futebol,0.99,1071
168,Now Sports,0.99,4884
170,A Statistic,0.99,6373
178,Oprah,0.99,6635
3304,Commercial 1,0.99,7941
172,The Real Problem,0.99,11650
3310,Commercial 2,0.99,21211
2241,Bossa,0.99,29048
1086,Casinha Feliz,0.99,32287
246,Mateus Enter,0.99,33149


In [5]:
%%sql
-- This query finds all customers who live in Brazil.
SELECT 
    CustomerID, 
    FirstName,
    LastName,
    Country
FROM 
    customer
WHERE Country = "Brazil";

 * mysql+pymysql://root:***@localhost:3306/chinook
5 rows affected.


CustomerID,FirstName,LastName,Country
1,Luís,Gonçalves,Brazil
10,Eduardo,Martins,Brazil
11,Alexandre,Rocha,Brazil
12,Roberto,Almeida,Brazil
13,Fernanda,Ramos,Brazil


# 2: Simple Aggregations (SUM, AVG, COUNT)

In [6]:
%%sql
    /* This query finds out how much money the store has made in total sales, the average amount spent per invoice, 
    and the total number of individual orders placed.*/
SELECT 
    SUM(Total) AS 'Total sales',
    AVG(Total) AS 'Average Sales',
    COUNT(InvoiceID) AS 'Orders placed'
FROM 
    invoice;

 * mysql+pymysql://root:***@localhost:3306/chinook
1 rows affected.


Total sales,Average Sales,Orders placed
2328.60,5.651942,412


In [7]:
%%sql
   /* This query calculates the total revenue (SUM), the average invoice amount (AVG), and the 
    highest single purchase (MAX) out of all invoices billed to the USA (BillingCountry = 'USA').*/
SELECT
    BillingCountry,
    SUM(Total) AS 'Total_revenue',
    AVG(Total) AS 'Invoice amount',
    MAX(Total) AS 'Highest_Purchase'
FROM 
    invoice
WHERE BillingCountry = "USA";


 * mysql+pymysql://root:***@localhost:3306/chinook
1 rows affected.


BillingCountry,Total_revenue,Invoice amount,Highest_Purchase
USA,523.06,5.747912,23.86


# 3: Grouping Data (GROUP BY)

In [8]:
%%sql
    -- This query calculates the total revenue generated by each country, ordered from our highest-earning market down to our lowest.
SELECT 
    BillingCountry,
    SUM(Total) AS 'Total_revenue',
    COUNT(InvoiceID) AS 'Orders made'
FROM 
    Invoice
GROUP BY 
    BillingCountry
ORDER BY
    Total_revenue DESC;

 * mysql+pymysql://root:***@localhost:3306/chinook
24 rows affected.


BillingCountry,Total_revenue,Orders made
USA,523.06,91
Canada,303.96,56
France,195.10,35
Brazil,190.10,35
Germany,156.48,28
United Kingdom,112.86,21
Czech Republic,90.24,14
Portugal,77.24,14
India,75.26,13
Chile,46.62,7


In [9]:
%%sql
-- This query calculates the total revenue generated by each individual city in Canada
SELECT
    BillingCountry,
    BillingCity,
    SUM(Total) AS 'Total_revenue'
FROM 
    invoice
WHERE 
    BillingCountry = "Canada"
GROUP BY 
    billingcity
ORDER BY 
    Total_revenue DESC;

 * mysql+pymysql://root:***@localhost:3306/chinook
8 rows affected.


BillingCountry,BillingCity,Total_revenue
Canada,Montréal,39.62
Canada,Vancouver,38.62
Canada,Edmonton,37.62
Canada,Halifax,37.62
Canada,Yellowknife,37.62
Canada,Toronto,37.62
Canada,Ottawa,37.62
Canada,Winnipeg,37.62


# 4: Filtering Groups (HAVING)

In [10]:
%%sql
    -- This query finds only the top-tier customers who have spent a combined total of more than $45 across all of their orders.
SELECT 
    CustomerID,
    SUM(Total) AS 'Total_Revenue',
    COUNT(InvoiceID) AS 'Orders made'
FROM 
    Invoice
GROUP BY
    CustomerID
HAVING 
    Total_revenue > 45
ORDER BY 
    Total_Revenue DESC;

 * mysql+pymysql://root:***@localhost:3306/chinook
5 rows affected.


CustomerID,Total_Revenue,Orders made
6,49.62,7
26,47.62,7
57,46.62,7
45,45.62,7
46,45.62,7


In [11]:
%%sql
   /* This query finds the corporate accounts or bulk-buyers by grouping the records in the Invoice 
    table by CustomerId to find customers who have placed more than 6 individual orders total.*/
SELECT 
    CustomerID,
    Count(InvoiceID) AS 'orders'
FROM 
    invoice
GROUP BY
    CustomerID
HAVING 
    orders > 6
ORDER BY orders DESC;

 * mysql+pymysql://root:***@localhost:3306/chinook
58 rows affected.


CustomerID,orders
1,7
2,7
3,7
4,7
5,7
6,7
7,7
8,7
9,7
10,7


# Complex Challenge 1: The High-Value Global Market Analysis
The Business Scenario: The marketing department wants to identify which non-US countries are our most reliable, high-value markets. They want to ignore small, one-off purchases and focus purely on countries where the average transaction size is solid, and the total revenue is substantial.

Your Task: Write a single query that satisfies these conditions:

Look at all invoices outside the USA (BillingCountry != 'USA').

Group the data by BillingCountry.

Calculate three metrics for each country: Total Revenue, Average Invoice Amount, and the Total Number of Orders.

Filter the final grouped results to only show countries where the Total Revenue is greater than $40.

Sort the output so that the country with the highest average invoice amount appears first.

In [71]:
%%sql
SELECT
    BillingCountry,
    SUM(Total) AS 'Total_Revenue',
    AVG(Total) AS 'Average invoice amount',
    COUNT(InvoiceID) AS 'Total Orders'
FROM
    invoice
WHERE 
    BillingCountry != 'USA'  
GROUP BY 
    BillingCountry
HAVING 
    SUM(Total) > 40 
ORDER BY 
    AVG(Total) DESC;

 * mysql+pymysql://root:***@localhost:3306/chinook
14 rows affected.


BillingCountry,Total_Revenue,Average invoice amount,Total Orders
Chile,46.62,6.660000,7
Ireland,45.62,6.517143,7
Hungary,45.62,6.517143,7
Czech Republic,90.24,6.445714,14
Austria,42.62,6.088571,7
Finland,41.62,5.945714,7
Netherlands,40.62,5.802857,7
India,75.26,5.789231,13
Germany,156.48,5.588571,28
France,195.10,5.574286,35


# Complex Challenge 2: Employee Performance Check (Fraud & Outlier Detection)
The Business Scenario: Management wants to check the consistency of our order values. They are looking for individual customer profiles where a single, massive order might be distorting that customer's lifetime value statistics.

Your Task: Write a single query to flag these specific profiles:

Group all transactions in the Invoice table by CustomerId.

Calculate the Highest Single Purchase (MAX) and the Average Purchase (AVG) for each customer.

Filter the groups using a HAVING clause to only include customers whose highest single purchase is more than 3 times larger than their average purchase.

Sort the results from the highest maximum purchase down to the lowest.

In [72]:
%%sql
SELECT
    CustomerID,
    AVG(Total) AS 'Average Purchase',
    MAX(Total) AS 'Highest Single Purchase'
FROM invoice
GROUP BY CustomerID
HAVING  MAX(Total) > 3*AVG(Total)
ORDER BY MAX(Total) DESC;

 * mysql+pymysql://root:***@localhost:3306/chinook
6 rows affected.


CustomerID,Average Purchase,Highest Single Purchase
6,7.088571,25.86
26,6.802857,23.86
45,6.517143,21.86
46,6.517143,21.86
7,6.088571,18.86
25,6.088571,18.86


# Advanced SQL Concepts

# **1.JOINS(Inner Join,Left Join)**

In [12]:
%%sql
    -- This query links Customers with their dedicated Support Representative (Employee) to see who manages who
SELECT 
    c.CustomerId,
    CONCAT(c.FirstName, ' ', c.LastName) AS 'Customer Name',
    c.City AS 'Customer City',
    CONCAT(e.FirstName, ' ', e.LastName) AS 'Support Rep Name',
    e.Title AS 'Rep Title'
FROM 
    Customer c
INNER JOIN 
    Employee e ON e.EmployeeId = c.SupportRepId
ORDER BY 
    c.CustomerId
    LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/chinook
10 rows affected.


CustomerId,Customer Name,Customer City,Support Rep Name,Rep Title
1,Luís Gonçalves,São José dos Campos,Jane Peacock,Sales Support Agent
2,Leonie Köhler,Stuttgart,Steve Johnson,Sales Support Agent
3,François Tremblay,Montréal,Jane Peacock,Sales Support Agent
4,Bjørn Hansen,Oslo,Margaret Park,Sales Support Agent
5,František Wichterlová,Prague,Margaret Park,Sales Support Agent
6,Helena Holý,Prague,Steve Johnson,Sales Support Agent
7,Astrid Gruber,Vienne,Steve Johnson,Sales Support Agent
8,Daan Peeters,Brussels,Margaret Park,Sales Support Agent
9,Kara Nielsen,Copenhagen,Margaret Park,Sales Support Agent
10,Eduardo Martins,São Paulo,Margaret Park,Sales Support Agent


In [13]:
%%sql 
    -- This query lists the top 10 longest tracks along with their Album Title and the Artist Name.
SELECT 
    t.Name AS 'Track Name',
    t.milliseconds,
    a.Title AS 'Album Title',
    r.Name AS 'Artist Name'
FROM
    track t
INNER JOIN 
   album a ON t.albumid = a.albumid
INNER JOIN
    artist r ON r.artistID = a.artistID
ORDER BY 
    t.Milliseconds DESC
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/chinook
10 rows affected.


Track Name,milliseconds,Album Title,Artist Name
Occupation / Precipice,5286953,"Battlestar Galactica, Season 3",Battlestar Galactica
Through a Looking Glass,5088838,"Lost, Season 3",Lost
"Greetings from Earth, Pt. 1",2960293,"Battlestar Galactica (Classic), Season 1",Battlestar Galactica (Classic)
The Man With Nine Lives,2956998,"Battlestar Galactica (Classic), Season 1",Battlestar Galactica (Classic)
"Battlestar Galactica, Pt. 2",2956081,"Battlestar Galactica (Classic), Season 1",Battlestar Galactica (Classic)
"Battlestar Galactica, Pt. 1",2952702,"Battlestar Galactica (Classic), Season 1",Battlestar Galactica (Classic)
Murder On the Rising Star,2935894,"Battlestar Galactica (Classic), Season 1",Battlestar Galactica (Classic)
"Battlestar Galactica, Pt. 3",2927802,"Battlestar Galactica (Classic), Season 1",Battlestar Galactica (Classic)
Take the Celestra,2927677,"Battlestar Galactica (Classic), Season 1",Battlestar Galactica (Classic)
Fire In Space,2926593,"Battlestar Galactica (Classic), Season 1",Battlestar Galactica (Classic)


In [14]:
%%sql
    -- This query finds "dormant" customers who have registered in the system  but have never placed a single order.
SELECT
    c.CustomerID,
    CONCAT(c.FirstName, ' ',c.LastName) AS 'Customer Name',
    c.email
FROM 
    customer c
LEFT JOIN 
    invoice i ON i.CustomerID = c.CustomerId
WHERE i.InvoiceID IS NULL;

 * mysql+pymysql://root:***@localhost:3306/chinook
0 rows affected.


CustomerID,Customer Name,email


In [15]:
%%sql
   /* This query verifies that every music genre listed in the database actually
    has tracks assigned to it, or if there are empty genres sitting in the catalog.*/
SELECT 
    g.GenreId,
    g.name AS 'Music Genre',
    COUNT(t.trackid) AS 'no of tracks'
FROM 
    track t
RIGHT JOIN 
   genre g ON t.genreid = g.genreid
GROUP BY 
    g.GenreId,
    g.name 
ORDER BY 'no of tracks' DESC
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/chinook
10 rows affected.


GenreId,Music Genre,no of tracks
1,Rock,1297
2,Jazz,130
3,Metal,374
4,Alternative & Punk,332
5,Rock And Roll,12
6,Blues,81
7,Latin,579
8,Reggae,58
9,Pop,48
10,Soundtrack,43


In [16]:
%%sql
    /* This query shows a list of individual item purchases, showing the track name, 
    the media type format (e.g., MPEG audio file, AAC audio file), and the price paid on the invoice.*/
SELECT 
    t.Name As 'track name',
    m.Name AS 'Media type',
    i.UnitPrice
FROM 
    track t
INNER JOIN 
    mediatype m ON m.mediatypeid = t.mediatypeid
INNER JOIN 
    invoiceline i ON i.trackid = t.trackid
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/chinook
10 rows affected.


track name,Media type,UnitPrice
Balls to the Wall,Protected AAC audio file,0.99
Restless and Wild,Protected AAC audio file,0.99
Put The Finger On You,MPEG audio file,0.99
Inject The Venom,MPEG audio file,0.99
Evil Walks,MPEG audio file,0.99
Breaking The Rules,MPEG audio file,0.99
Dog Eat Dog,MPEG audio file,0.99
Overdose,MPEG audio file,0.99
Love In An Elevator,MPEG audio file,0.99
Janie's Got A Gun,MPEG audio file,0.99


# **2. Subqueries**

In [17]:
%%sql
    -- This query finds all tracks that are longer than the overall average length of all tracks in the store.
SELECT 
    TrackId, 
    Name, 
    Milliseconds / 1000 AS 'Duration (Seconds)'
FROM 
    Track
WHERE 
    Milliseconds > (
        SELECT AVG(Milliseconds) FROM Track
    )
ORDER BY 
    Milliseconds DESC
    LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/chinook
10 rows affected.


TrackId,Name,Duration (Seconds)
2820,Occupation / Precipice,5286.9530
3224,Through a Looking Glass,5088.8380
3244,"Greetings from Earth, Pt. 1",2960.2930
3242,The Man With Nine Lives,2956.9980
3227,"Battlestar Galactica, Pt. 2",2956.0810
3226,"Battlestar Galactica, Pt. 1",2952.7020
3243,Murder On the Rising Star,2935.8940
3228,"Battlestar Galactica, Pt. 3",2927.8020
3248,Take the Celestra,2927.6770
3239,Fire In Space,2926.5930


In [18]:
%%sql
    -- This query finds all customers who have bought at least one track from the 'Rock' genre.
SELECT 
    CustomerId,
    FirstName,
    LastName,
    Email
FROM 
    Customer
WHERE 
    CustomerId IN (
        SELECT DISTINCT 
            i.CustomerId
        FROM 
            Invoice i
        INNER JOIN 
            InvoiceLine il ON i.InvoiceId = il.InvoiceId
        INNER JOIN 
            Track t ON il.TrackId = t.TrackId
        INNER JOIN 
            Genre g ON t.GenreId = g.GenreId
        WHERE 
            g.Name = 'Rock'
    );

 * mysql+pymysql://root:***@localhost:3306/chinook
59 rows affected.


CustomerId,FirstName,LastName,Email
47,Lucas,Mancini,lucas.mancini@yahoo.it
2,Leonie,Köhler,leonekohler@surfeu.de
33,Ellie,Sullivan,ellie.sullivan@shaw.ca
13,Fernanda,Ramos,fernadaramos4@uol.com.br
4,Bjørn,Hansen,bjorn.hansen@yahoo.no
8,Daan,Peeters,daan_peeters@apple.be
53,Phil,Hughes,phil.hughes@gmail.com
22,Heather,Leacock,hleacock@gmail.com
14,Mark,Philips,mphilips12@shaw.ca
42,Wyatt,Girard,wyatt.girard@yahoo.fr


In [19]:
%%sql
    -- This query finds the average total spend per customer.
SELECT 
    AVG(CustomerTotalSpend) AS 'Average Spend Per Customer'
FROM (
    -- Calculates total spend per individual customer
    SELECT 
        CustomerId,
        SUM(Total) AS CustomerTotalSpend
    FROM 
        Invoice
    GROUP BY 
        CustomerId
) AS CustomerSpends;

 * mysql+pymysql://root:***@localhost:3306/chinook
1 rows affected.


Average Spend Per Customer
39.467797


# **3. Window Functions**

In [20]:
%%sql
    /*This query displays the track Name, AlbumId, Milliseconds, and assign a rank
    to each track based on its duration within its album (longest track = Rank 1).*/
SELECT 
    AlbumId,
    Name AS 'Track Name',
    Milliseconds,
    ROW_NUMBER() OVER (
        PARTITION BY AlbumId 
        ORDER BY Milliseconds DESC
    ) AS Track_Length_Rank
FROM 
    Track
ORDER BY 
    AlbumId, 
    Track_Length_Rank;

 * mysql+pymysql://root:***@localhost:3306/chinook
3503 rows affected.


AlbumId,Track Name,Milliseconds,Track_Length_Rank
1,For Those About To Rock (We Salute You),343719,1
1,Spellbound,270863,2
1,Evil Walks,263497,3
1,Breaking The Rules,263288,4
1,Let's Get It Up,233926,5
1,Inject The Venom,210834,6
1,Night Of The Long Knives,205688,7
1,Put The Finger On You,205662,8
1,Snowballed,203102,9
1,C.O.D.,199836,10


In [21]:
%%sql
    -- This query ranks all customers by their overall total spending across the entire business using DENSE_RANK()
SELECT 
    CustomerId,
    SUM(Total) AS Total_Spent,
    DENSE_RANK() OVER (
        ORDER BY SUM(Total) DESC
    ) AS Overall_Rank
FROM 
    Invoice
GROUP BY 
    CustomerId
ORDER BY 
    Overall_Rank;

 * mysql+pymysql://root:***@localhost:3306/chinook
59 rows affected.


CustomerId,Total_Spent,Overall_Rank
6,49.62,1
26,47.62,2
57,46.62,3
46,45.62,4
45,45.62,4
37,43.62,5
28,43.62,5
24,43.62,5
25,42.62,6
7,42.62,6


# Business Problem Solving

# **1: Top-Performing Products & Customers**

In [52]:
%%sql
SELECT 
    c.CustomerId,
    CONCAT(c.FirstName, ' ', c.LastName) AS 'Customer Name',
    c.Country,
    COUNT(i.InvoiceId) AS 'Total Orders Placed',
    ROUND(SUM(i.Total), 2) AS 'Lifetime Spend ($)'
FROM 
    Customer c
INNER JOIN 
    Invoice i ON c.CustomerId = i.CustomerId
GROUP BY 
    c.CustomerId, 
    c.FirstName, 
    c.LastName, 
    c.Country
ORDER BY 
    `Lifetime Spend ($)` DESC
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/chinook
5 rows affected.


CustomerId,Customer Name,Country,Total Orders Placed,Lifetime Spend ($)
6,Helena Holý,Czech Republic,7,49.62
26,Richard Cunningham,USA,7,47.62
57,Luis Rojas,Chile,7,46.62
45,Ladislav Kovács,Hungary,7,45.62
46,Hugh O'Reilly,Ireland,7,45.62


# **2: Revenue Trends Over Time**

In [53]:
%%sql
SELECT 
    YEAR(InvoiceDate) AS 'Sales Year',
    MONTH(InvoiceDate) AS 'Sales Month',
    COUNT(InvoiceId) AS 'Monthly Orders',
    ROUND(SUM(Total), 2) AS 'Monthly Revenue ($)'
FROM 
    Invoice
GROUP BY 
    YEAR(InvoiceDate), 
    MONTH(InvoiceDate)
ORDER BY 
    'Sales Year' ASC, 
    'Sales Month' ASC;

 * mysql+pymysql://root:***@localhost:3306/chinook
60 rows affected.


Sales Year,Sales Month,Monthly Orders,Monthly Revenue ($)
2021,1,6,35.64
2021,2,7,37.62
2021,3,7,37.62
2021,4,7,37.62
2021,5,7,37.62
2021,6,7,37.62
2021,7,7,37.62
2021,8,7,37.62
2021,9,7,37.62
2021,10,7,37.62


# **3: Top-Performing Tracks by Revenue**

In [54]:
%%sql
SELECT 
    t.TrackId,
    t.Name AS 'Track Name',
    ROUND(SUM(il.UnitPrice * il.Quantity), 2) AS Total_Sales_Amount
FROM 
    Track t
INNER JOIN 
    InvoiceLine il ON t.TrackId = il.TrackId
GROUP BY 
    t.TrackId, 
    t.Name
ORDER BY 
    Total_Sales_Amount DESC
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/chinook
5 rows affected.


TrackId,Track Name,Total_Sales_Amount
2832,The Woman King,3.98
2868,Walkabout,3.98
2850,The Fix,3.98
3223,How to Stop an Exploding Man,3.98
3214,Phyllis's Wedding,3.98


# **4: Yearly Country Growth Trend**

In [55]:
%%sql
SELECT 
    YEAR(InvoiceDate) AS Sales_Year,
    ROUND(SUM(Total), 2) AS Total_Revenue
FROM 
    Invoice
WHERE 
    BillingCountry = 'USA'
GROUP BY 
    YEAR(InvoiceDate)
ORDER BY 
    Sales_Year ASC;

 * mysql+pymysql://root:***@localhost:3306/chinook
5 rows affected.


Sales_Year,Total_Revenue
2021,103.95
2022,102.98
2023,103.01
2024,127.98
2025,85.14


# Document explaining queries and insights

**1. Schema Exploration & Database Structure**
    
Database Setup & Table Listing
- Query: Connected via %sql mysql+pymysql and executed SHOW TABLES; to review the database contents.  
**Findings:** The database consists of 11 tables: album, artist, customer, employee, genre, invoice, invoiceline, mediatype, playlist, playlisttrack, and track.  

Table Schema & Primary/Foreign Keys
- Queries: Executed DESCRIBE statements across all tables, followed by a metadata query on information_schema.KEY_COLUMN_USAGE.  
- Key Structural Relationships:
    - Customer & Sales: customer.CustomerId $\rightarrow$ invoice.CustomerId $\rightarrow$ invoiceline.InvoiceId.  
    - Catalog & Media: artist.ArtistId $\rightarrow$ album.ArtistId $\rightarrow$ track.AlbumId.  
    - Categorization: track connects to genre via GenreId and to mediatype via MediaTypeId.  
    - Playlists: Many-to-many relationship managed by the bridge table playlisttrack (PlaylistId, TrackId).  
    - Internal HR: employee.ReportsTo self-references employee.EmployeeId, while customer.SupportRepId maps support representatives to customers.

Table Sizing Summary
- Query: Joined information_schema.TABLES and information_schema.COLUMNS to evaluate row and column counts.  
- **Insights:**
    - Largest Data Tables: playlisttrack (8,715 rows) and track (3,483 rows) represent the core media library.  
    - Sales Volume: 412 total invoice records and 2,240 invoiceline records.  
    - Customer Base: 59 registered customers served by 8 employees.  

**2. Filtering, Sorting, and Aggregations**
    
Filtered Catalog Audits
- Track Length (Album 1): Identified 4 tracks on Album 1 exceeding 4 minutes ($240,000\text{ ms}$), with "For Those About To Rock (We Salute You)" being the longest ($343,719\text{ ms}$). 
- Budget Music Promotion: Filtered for tracks priced at $\$0.99$ under 3 minutes ($180,000\text{ ms}$). Shortest tracks include "É Uma Partida De Futebol" ($1,071\text{ ms}$) and short commercial spot tracks.  
- Demographic Target: Identified 5 customers based in Brazil (Luís Gonçalves, Eduardo Martins, Alexandre Rocha, Roberto Almeida, Fernanda Ramos) for localized marketing.  

Overall Sales Performance Metrics
- Query: Calculated SUM(Total), AVG(Total), and COUNT(InvoiceID) on the invoice table.  
**Key Financial Metrics:**
    - Total Revenue: $\$2,328.60$  
    - Total Orders Placed: 412 invoices  
    - Average Invoice Value: $\$5.65$  
    
Market Performance: USA
- Query: Aggregated sales metrics specifically for BillingCountry = 'USA'.  
- USA Highlights: Generated $\$523.06$ across 91 orders with an average spend of $\$5.75$ per order and a single maximum purchase of $\$23.86$.  

**3. Grouped Revenue & High-Value Segment Analysis**

**Regional Revenue Breakdown**
- Query: Grouped invoices by BillingCountry and ordered by total revenue.  
- Top Geographic Markets:
    - USA: $\$523.06$ (91 orders)  
    - Canada: $\$303.96$ (56 orders) 
    - France: $\$195.10$ (35 orders)  
    - Brazil: $\$190.10$ (35 orders) 
    - Germany: $\$156.48$ (28 orders)  

**Canadian Breakdown by City**
- Query: Evaluated revenue across Canadian cities.  
- **Insights:** Revenue is evenly distributed across 8 Canadian cities, with Montréal leading slightly at $\$39.62$, followed by Vancouver at $\$38.62$, and six other cities (Edmonton, Halifax, Yellowknife, Toronto, Ottawa, Winnipeg) tied at $\$37.62$ each.  

**Customer Spending Tier Analysis**
- Top Spenders (HAVING Total_Revenue > 45): Identified 5 top-tier customers who spent over $\$45$ total across their order histories, led by CustomerID 6 ($\$49.62$) and CustomerID 26 ($\$47.62$).  
- Order Frequency: Every single registered customer (IDs 1 through 58) has completed exactly 7 orders in the system.  

**4. Complex Challenge: High-Value Global Markets (Excluding USA)**
- Query Scenario: Analyzed non-US markets generating over $\$40$ in total revenue, sorted by average invoice value.  
- **Key Strategic Insights:** 
- Highest Average Order Value (AOV): Chile leads non-US markets with an average transaction value of $\$6.66$, followed closely by Ireland ($\$6.52$) and Hungary ($\$6.52$). 
- High Volume vs. High Value: While Canada ($\$303.96$) and France ($\$195.10$) dominate total revenue output due to higher transaction counts, smaller markets like Chile, Ireland, Czech Republic ($\$6.45\text{ AOV}$), and Hungary demonstrate higher average spend per transaction.